File: system-test/dotnet/Dsl.Core/UseCase/External/Erp/ErpDsl.cs
using Driver.Port.External.Erp;
using Dsl.Core.External.Erp.UseCases;
using Dsl.Core.Shared;


namespace Dsl.Core.External.Erp;

public class ErpDsl : IDisposable, IAsyncDisposable
{
    private readonly IErpDriver _driver;
    private readonly UseCaseContext _context;
    private bool _disposed;

    public ErpDsl(IErpDriver driver, UseCaseContext context)
    {
        _driver = driver;
        _context = context;
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    public async ValueTask DisposeAsync()
    {
        await _driver.DisposeAsync();
        _disposed = true;
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (_disposed) return;
        if (disposing)
            _driver?.Dispose();
        _disposed = true;
    }

    public GoToErp GoToErp() => new(_driver, _context);

    public ReturnsProduct ReturnsProduct() => new(_driver, _context);

    public ReturnsPromotion ReturnsPromotion() => new(_driver, _context);

    public GetProduct GetProduct() => new(_driver, _context);
}

File: system-test/java/src/main/java/com/optivem/shop/testkit/core/usecase/external/erp/ErpDsl.java
package com.optivem.shop.testkit.core.usecase.external.erp;

import com.optivem.shop.testkit.driver.port.external.erp.ErpDriver;
import com.optivem.shop.testkit.core.usecase.external.erp.usecases.GetProduct;
import com.optivem.shop.testkit.core.usecase.external.erp.usecases.GoToErp;
import com.optivem.shop.testkit.core.usecase.external.erp.usecases.ReturnsProduct;
import com.optivem.shop.testkit.core.usecase.external.erp.usecases.ReturnsPromotion;
import com.optivem.shop.testkit.common.Closer;
import com.optivem.shop.testkit.core.shared.UseCaseContext;

public class ErpDsl implements AutoCloseable {
    protected final ErpDriver driver;
    protected final UseCaseContext context;

    public ErpDsl(ErpDriver driver, UseCaseContext context) {
        this.driver = driver;
        this.context = context;
    }

    @Override
    public void close() {
        Closer.close(driver);
    }

    public GoToErp goToErp() {
        return new GoToErp(driver, context);
    }

    public ReturnsProduct returnsProduct() {
        return new ReturnsProduct(driver, context);
    }

    public GetProduct getProduct() {
        return new GetProduct(driver, context);
    }

    public ReturnsPromotion returnsPromotion() {
        return new ReturnsPromotion(driver, context);
    }
}
